netty: Handle shutdown and failures during negotiator

ejona86 opened the pull request:
NettyClientTransport needs to call close() on the Channel directly
instead of sending a message, since the message would typically be
delayed until negotiation completes.

The closeFuture() closes too early to be helpful, which is very
unfortunate. Using it squelches the negotiator's error handling. We now
rely on the handlers to report shutdown without any back-up. The
handlers error handling has matured, so maybe this is okay.

------

This work was initially triggered by useless error messages if using a Proxy failed. I _thought_ I had already submitted this. But then it bit me again when trying to get Conscrypt working, so I finished up the old commit.

carl-mastrangelo commented on netty/src/main/java/io/grpc/netty/WriteQueue.java:
```diff
@@ -100,6 +100,18 @@ ChannelFuture enqueue(QueuedCommand command, ChannelPromise promise, boolean flu
+   * processed in-order with writes.
+   */
+  void enqueue(Runnable runnable, boolean flush) {
+    queue.add(new RunnableCommand(runnable));
```
you might consider chaining this with the above enqueue and just use a voidPromise

ejona86 replied:
Hmm... I'd also need to "support" the setter `QueuedCommand.promise(ChannelPromise)`. It feels about the same to me. I'm going to leave it as-is; this is a pretty straight-forward method and would be easy to change in the future if it became a larger problem.

ejona86 commented:
Rebased to absorb conflict with #3351. I changed `NettyCLientTransport.shutdownNow()` to pass `reason` to `lifecycleManager` instead of creating its own `Status`; that preserves the behavior after #3351.
